crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+//! Common types for implementing federation authorization
+
+use headers::{authorization::Credentials, HeaderValue};
+use ruma_common::{ServerName, ServerSigningKeyId};
+use tracing::debug;
+
+/// Typed representation of an `Authorization` header of scheme `X-Matrix`, as defined in the
+/// [Matrix Server-Server API][spec]. Includes an implementation of
+/// [`headers::authorization::Credentials`] for automatically handling the encoding and decoding
+/// when using a web framework that supports typed headers.
+///
+/// [spec]: https://spec.matrix.org/v1.3/server-server-api/#request-authentication
+#[non_exhaustive]
+pub struct XMatrix {
+    /// the server name of the sending server
+    pub origin: Box<ServerName>,
```

Comment:
```suggestion
    pub origin: OwnedServerName,
```

`Box<StringyType>` is out 😅  (applies below as well)